lib/eth/EthRpc.js

```diff
@@ -177,56 +179,178 @@ class EthRPC {
     return resultArray;
   }
 
-  estimateFee({ nBlocks }) {
-    return this.estimateGasPrice(nBlocks);
+  /**
+   * Estimates the fee for a transaction.
+   * @param {Object} params - The parameters for estimating the fee.
+   * @param {number} params.nBlocks - The number of blocks to consider for the estimation.
+   * @param {string} [params.txType='0'] - The type of transaction fee estimation. '2' for EIP-1559 style, '0' for legacy.
+   * @returns {Promise<Object|number>} The estimated fee. If txType is '2', returns an object with maxFee and priorityFee. Otherwise, returns a number.
+   */
+  async estimateFee({ nBlocks, txType = '0' }) {
+    switch (txType) {
+      case '2':
+        return { 
+          maxFee: await this.estimateMaxFee({}), 
+          priorityFee: await this.estimatePriorityFee({}) 
+        };
+      default:
+        return await this.estimateGasPrice(nBlocks);
+    }
   }
 
+  /**
+   * Estimates the fee.
+   * @param {Object} params - The parameters for estimating the fee.
+   * @param {Map} params.cache - The cache of block fees.
+   * @param {number} params.rpcEstimate - The RPC estimate of the fee.
+   * @param {number} [params.percentile=25] - The percentile to use for the estimation.
+   * @param {Function} params.getFees - The function to get the fees from the blocks.
+   * @returns {Promise<number>} The estimated fee.
+   */
+  async _estimateFee({ cache, rpcEstimate, percentile = 25, getFees }) {
+    const blocks = await this._getRecentBlocks({ cache });
+    const percentileFees = this._getPercentileFees({ percentile, blocks, cache, getFees });
+    if (!percentileFees.length) {
+      return rpcEstimate;
+    }
+    const estimate = this._calculateFeeEstimate(percentileFees);
+    return Math.max(estimate, rpcEstimate);
+  }
+
+  /**
+   * Estimates the gas price.
+   * @returns {Promise<number>} The estimated gas price.
+   */
   async estimateGasPrice() {
+    const rpcEstimate = parseInt(await this.web3.eth.getGasPrice());
+    return this._estimateFee({
+      cache: this.blockGasPriceCache,
+      rpcEstimate,
+      percentile: 25,
+    });
+  }
+
+  /**
+   * Estimates the maximum base fee.
+   * @param {Object} params - The parameters for estimating the maximum base fee.
+   * @param {number} [params.percentile=25] - The percentile to use for the estimation.
+   * @returns {Promise<number>} The estimated maximum base fee.
+   */
+  async estimateMaxFee({ percentile = 25 }) {
+    const lastBlock = await this.web3.eth.getBlock('latest');
+    // a blocks base fee is the lowest acceptied MaxFee from the last block. 
+    const rpcEstimate = lastBlock.baseFeePerGas ? lastBlock.baseFeePerGas : this.web3.utils.toWei('1', 'gwei');
+    const parsedRpcEstimate = parseInt(rpcEstimate);
+    const getFees = (_txs) => _txs
+      .filter((tx) => !!tx.maxFeePerGas)
+      .map((tx) => parseInt(tx.maxFeePerGas));
+    return this._estimateFee({
```

Comment:
I could be wrong here but I am wondering what the benefit of using the percentile is for establishing the max fee. Part of the benefits of EIP-1559 is establishing a more predictable fee market and I think part of that is due to being able to know exactly within what range the base fee will be any number of blocks in the future with out having to parse fee history. The important data to check would be how full the most recent blocks have been and let the max base fee be the outcome of that + the priority preference for the transaction. If blocks are full and the priority preference is high then you would aim for 2x the base fee (+ priority fee) for surety of not being priced out within the next 6 blocks. Priority preference being low and you would just use the current base fee + the priority fee. If blocks are underutilized then you have a greater certainty of inclusion in either case and can set the max fee accordingly.

In the end whatever is unused of the max fee is refunded to the user so this really boils down to the certainty of inclusion within the next few blocks and the users preference for fast/slow inclusion as it relates to the fee they are willing to pay. If it wasn't a bad UX I would say just set it to 2x base fee + priority fee. The downside is that every time a user signs a transaction it will appear that it will cost 2x more than it potentially could cost.